File: my-folder/0139-word-break/solution.cpp
class Solution {
public:
    int dp[301];
    bool solve(string s,int idx,set<string> st){
        if(idx>=s.size()) return true;
        if(dp[idx]!=-1) return dp[idx];
        string temp ="";
        for(int i=idx;i<s.size();i++){
            temp+=s[i];
            if(st.find(temp)!=st.end()){
                if(solve(s,i+1,st)) return dp[idx]=1;
            } 
        }
        return dp[idx]=0;
    }
    bool wordBreak(string s, vector<string>& wordDict) {
        memset(dp,-1,sizeof(dp));
        set<string> st;
        for(string word: wordDict) st.insert(word);
        return solve(s,0,st);
        
    }
};

File: my-folder/0139-word-break/solution.java

class Solution {
    int[] dp = new int[301];

    private boolean solve(Set<String> st, String s, int idx) {
        if (idx >= s.length()) {
            return true;
        }
        if (dp[idx] != 0) {
            return dp[idx] == 1;
        }

        for (int i = 1; i + idx <= s.length(); i++) {
            String temp = s.substring(idx, idx + i);
            if (st.contains(temp)) {
                if (solve(st, s, idx + i)) {
                    dp[idx] = 1;
                    return true;
                }
            }
        }

        dp[idx] = -1;
        return false;
    }

    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> st = new HashSet<>(wordDict);
        Arrays.fill(dp, 0);
        return solve(st, s, 0);
    }
}

